### ISSUE TITLE
Is OK to use cosine_similarity instead softmax for VISION x TEXT ?

### ISSUE THREAD

**Issue by XinyueZ** [facebookresearch/ImageBind#72]
Hey,

I just want to know if the cosine_similarity of sklearn can relplace the softmax.

Thanks

**Comment by XinyueZ** [comment]
yeah, agree with @DaNious, tbh, I did the `similarity` comparing with the `cosine_similarity`, and also get a proper result.

**Comment by bakachan19** [comment]
@XinyueZ, I am interested in computing the text x image similarity. But I am confused on how to do this.
Could you please share an example code?
Much appreciated.
Thank you!

**Comment by XinyueZ** [comment]
``` python
class SimilarityCalculator:
    def __init__(self, device):
        self.device = device

    def __call__(self, embeddings):
        raise NotImplementedError

    @staticmethod
    def create_instance(similarity_type, device):
        if similarity_type == SimilarityCalculatorType.COSINE:
            return CosineSimilarity(device)
        elif similarity_type == SimilarityCalculatorType.SOFTMAX:
            return SoftmaxSimilarity(device)
        else:
            raise ValueError(f"Unknown similarity type: {similarity_type}")


class CosineSimilarity(SimilarityCalculator):
    def __init__(self, device):
        super().__init__(device)

    def __call__(self, embeddings):
        preds = cosine_similarity(
            embeddings[ModalityType.VISION].cpu().numpy(),
            embeddings[ModalityType.TEXT].cpu().numpy(),
        )
        preds = torch.from_numpy(preds).to(self.device)
        print(
            preds.shape,
            "\n",
            preds,
            "max: ",
            preds.max(dim=0),
            "min: ",
            preds.min(dim=0),
        )
        return preds


class SoftmaxSimilarity(SimilarityCalculator):
    def __init__(self, device):
        super().__init__(device)

    def __call__(self, embeddings):
        preds = torch.softmax(
            embeddings[ModalityType.VISION] @ embeddings[ModalityType.TEXT].T,
            dim=0,
        )
        print(
            preds.shape,
            "\n",
            preds,
            "sum: ",
            preds.sum(dim=0),
            "max: ",
            preds.max(dim=0),
            "min: ",
            preds.min(dim=0),
        )
        return preds


    def _infer(self, vision, text, threshold=0.5): 
        inputs = {
            ModalityType.VISION: ImageBindClassifier._load_and_transform_vision_data_np(
                vision,
                self.device + ":1" if self.device == "cuda" else self.device,
            ),
            ModalityType.TEXT: ImageBindClassifier._load_and_transform_text(
                text,
                self.bpe_path,
                self.device + ":0" if self.device == "cuda" else self.device,
            ),
        }

        with torch.no_grad():
            embeddings = self.model(inputs)

        preds = self.similarity_calculator(embeddings) # softmax OR from sklearn.metrics.pairwise import cosine_similarity
        return ImageBindClassifier._filter_indices_for_each_column(
            preds,
            text,
            threshold,
        )
``` @bakachan19

**Comment by bakachan19** [comment]
Thank you so so much @XinyueZ.

**Comment by abhimanyu891998** [comment]
Hey @XinyueZ , I have been experimenting with imageBind for videos, I essentially extract the clips (5 secs), audio and subtitles from a video and I want them all in the same embedding space. I have tried this with the vanilla imagebind implementation on my 3080ti GPU with 24GB memory. However, I am facing an issue where the embedding generation takes too long. For a 8 minute video, it takes 40 minutes to generate the embeddings for video clips, text segments and audio clips (each corresponding to every 5 second segment of the video).

I was wondering if I there is a way to speed up the inference, or if you know of a way to quantize imageBind model to accelerate this process? Or maybe I am doing something wrong, just wanted your advice on it. Thanks!
